tools/c7n_gcp/c7n_gcp/resources/kms.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2019 Capital One Services, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+from c7n.utils import local_session
+from c7n_gcp.provider import resources
+from c7n_gcp.query import QueryResourceManager, TypeInfo, ChildResourceManager, ChildTypeInfo
+
+
+@resources.register('kms-keyring')
+class KmsKeyRing(QueryResourceManager):
+
+    class resource_type(TypeInfo):
+        service = 'cloudkms'
+        version = 'v1'
+        component = 'projects.locations.keyRings'
+        enum_spec = ('list', 'keyRings[]', None)
+        scope = None
+        id = 'name'
+
+        @staticmethod
+        def get(client, resource_info):
+            name = 'projects/{}/locations/{}/keyRings/{}' \
+                .format(resource_info['project_id'],
+                        resource_info['location'],
+                        resource_info['key_ring_id'])
+            return client.execute_command('get', {'name': name})
+
+    def get_resource_query(self):
+        if 'query' in self.data:
```

Comment:
just do a static list of locations inline to the module as a variable and query for each for the default. ie. only look at query if the user specifies. its unclear to me what happens in the current state if the user doesn't specify.. i assume it just breaks or doesn't return anything... ?

Reply:
It breaks in the current state as the API expects a location as a parameter.

Reply:
> query for each for the default
Could you clarify what exactly you mean by that? The way I currently get it is calling the API with each location from the list, which would actually make the same overhead as the initial solution when there was the location resource. The other way though might be referring to the variable put at `region` flag, what do you think?

Reply:
1. use a static list of regions in the module or package, to loop over and query. it saves the api call for each child resource to fetch the static list of parents. the tradeoff is a burden to keep the list up to date, but the change rate here is 1-2 times per year. ideally a functional test so we can detect drift on the value.

1. if the user passes a query with location information use that (note this could be a list of locations, check for either scalar location: us-central1or location: [] or multiple singletons values passed in query array.

1. if we get passed a region on cli, we should use that as a location, we're don't currently use region much with gcp provider, as many apis provide global reach, the only additional consideration here is if its inconsistent, so it should be documented in the resource class doc strings wrt to behavior (this resource is region specific and respects --region flag).

There's probably a longer consideration in general for this provider on respecting region flags, but thats a different discussion.

if you have a strong preference to go back to kms location parent vs the above, then i think its okay.

Reply:
In the latest version, I'm calling the API myself via `client.execute_query` to get the list of locations unless specified by the user. As for the default flag from cli, considering the lack of support in other GCP resources giving up on this idea right now. Working on querying by an array as well rather than expecting only a single value.

Reply:
Done.